### PR TITLE
Add compatibility with ember-qunit@5

### LINKED ISSUE #812: Compatibility with `ember-qunit@5`
`ember-qunit@5` is (finally 😩) moving along, it brings a few major changes:

* Drop support for older Node versions
* Remove usage of APIs that are not Embroider compatible
  * `{{content-for` usage has been replaced by having the `tests/index.html` directly include our expected DOM content
  * `app.import` has been removed in favor of leveraging imports along with ember-auto-import / Embroider to process them
* Drop direct dependency on `qunit` and `@ember/test-helpers`; moving them to be `peerDependencies` (along with some helpful validations and error messages to help guide folks down the path)

The combination of these things results in a somewhat fundamental problem: tooling can no longer depend on the `QUnit` global to be defined. This is the primary issue that we need to determine how to resolve for `qunit-dom`.

Specifically, at the moment, we directly `app.import` into the `test-support.js` file:

https://github.com/simplabs/qunit-dom/blob/258f8b3a5b0d6f09364d26fec66d454bf3d471e4/index.js#L12-L22

Up until now, this has worked just fine. However, since `QUnit` itself is not directly "splatted" into the `test-support.js` file (it is within a module instead) we are no longer guaranteed that `window.QUnit` will be defined when we call `QUnit.assert.dom = ...` here:

https://github.com/simplabs/qunit-dom/blob/258f8b3a5b0d6f09364d26fec66d454bf3d471e4/lib/qunit-dom.ts#L11

### PR DESCRIPTION
**Pull request by rwjblue**
In order to add support for `ember-qunit@5` the primary Ember project usage pattern changes a bit:

```js
// tests/test-helper.js
import * as QUnit from 'qunit';
import { setup } from 'qunit-dom';
//...snip...
setup(QUnit.assert);
```

No change is required for ember-qunit < 5 (this should not be a breaking change).

Closes #812

### CODE REVIEW

**Comment by scalvert** [comment]
Looks good so far. Thanks for implementing explicit calls vs. side-effect imports!

**Comment by scalvert** [comment]
What's the status of this PR, @rwjblue? Can we publish for review, or are you still poking at it?

**Comment by rwjblue** [comment]
> What's the status of this PR, @rwjblue? Can we publish for review, or are you still poking at it?

I still have a few TODO's that need to be completed. 

> Also, publishing ember-qunit, @ember/test-helpers, and others will require a bit of coordination. Do we have a parent issue that determines the correct order of rollout for all these releases, other than Rob's head? :)

Good idea! Created https://github.com/ember-cli/ember-cli/issues/9335 with the TODOs.

File: tsconfig.json
Comment (rwjblue) [inline tsconfig.json:13]: Can be removed once we land #820

**Comment by scalvert** [comment]
Just did a rebase to fix merge conflicts.

**Comment by rwjblue** [comment]
https://github.com/rwjblue/--ember-qunit-5-sample-app is a demo app that uses the "whole new setup" and is passing.

**Comment by scalvert** [comment]
In terms of the API we expose to users, I'd prefer that we use the pattern that requires the hosting project to pass in `QUnit` to extend.

```js
import { setup } from "qunit-dom"; 

// ...

setup(QUnit);
```

This will not require any special configuration from `qunit-dom` itself, in addition to not making any assumptions about the consumer's configuration of their project. 

Thoughts?

**Comment by rwjblue** [comment]
> In terms of the API we expose to users, I'd prefer that we use the pattern that requires the hosting project to pass in QUnit to extend.

Ya, agreed. That's what the current implementation does (though I need to push a commit to the demo repo above with that change in it).

**Comment by scalvert** [comment]
In the PR description, these two are the same thing, no?
> Document a new API (still somewhat TBD, but probably import "qunit-dom" or import { setup } from "qunit-dom"; setup(QUnit);
> Decide if setup should accept QUnit argument or directly import it?

**Comment by scalvert** [comment]
@rwjblue I updated the description to better reflect the actual final work for this PR.

**Comment by scalvert** [comment]
@Turbo87 I think we've managed to get this into a decent state. The only thing we haven't done here is add automated tests that test the Ember variant with `ember-qunit` > 5, but @rwjblue has a repo that he's been using to manually test. How do we think we should handle this?

File: .github/workflows/ci.yml
Comment (Turbo87) [inline .github/workflows/ci.yml:44]: I think I'd prefer to use a `matrix` "strategy" here and have dedicated jobs for each of the ember-try scenarios

File: index.js
Comment (Turbo87) [inline index.js:50]: what is the purpose of this `this.preprocessJs()` call? might be good to add a code comment :)

File: ember-cli-build.js
Comment (Turbo87) [inline ember-cli-build.js:11]: I'm wondering if it would be better to use a `USE_EMBROIDER` environment variable for this and unconditionally add the dev deps to package.json, instead of using ember-try scenarios. That would allow dependabot to update them, and would not immediately break our CI if embroider releases a version that is no longer compatible.

File: package.json
Comment (Turbo87) [inline package.json:44]: seems a bit unfortunate that any non-ember user will now have to install webpack, but I don't know how to solve this any other way either. having a `ember-qunit-dom` package just to integrate things isn't really great either.

File: .github/workflows/ci.yml
Comment (rwjblue) [reply .github/workflows/ci.yml:44]: Sounds good. I was on the fence about this.

File: package.json
Comment (rwjblue) [reply package.json:44]: Now that I have the try scenarios setup, I'll try a few other permutations to see if we can avoid it.

Comment (rwjblue) [reply package.json]: just confirmed, and removed

File: ember-cli-build.js
Comment (rwjblue) [reply ember-cli-build.js]: Done!

File: index.js
Comment (rwjblue) [reply index.js:59]: Updated!

File: .github/workflows/ci.yml
Comment (rwjblue) [reply .github/workflows/ci.yml]: Done!

File: .github/workflows/ci.yml
Comment (Turbo87) [inline .github/workflows/ci.yml:51]: I assume this is a typo? 😅

File: package.json
Comment (Turbo87) [inline package.json:31]:
```suggestion
```

I guess we don't need this anymore :)

File: tests/dummy/config/ember-try.js
Comment (Turbo87) [inline tests/dummy/config/ember-try.js:9]:
```suggestion
        // this is currently the version in package.json, so no dependency changes are needed
```

**Review by Turbo87 — APPROVED** [review]
LGTM after these minor comments are resolved. nice work! 👍

**Pull request merged**
